### DISCUSSION TITLE
Enhanced `var` pattern

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/csharplang#7181] [Language Ideas] [answered]
Today C# allows us using `var` pattern to get result from an expression like:

```csharp
if (expr is { Property: var p, _field: var f })
    // ...
```

Sometimes, a property or field may be a ref return rather than a value return. Here may exist a need for defining inlining variables with ref using `var` pattern like:

```csharp
// properties
if (expr is
{
    RefProperty: ref var r,
    RefProperty2: ref readonly var s,
    RefReadOnlyProperty: ref readonly var t
})
    // ...

// indexers
if (span is [ref var a, ref readonly var b, ..])
    // ...

// methods
if (Method() is ref var r)
    // ...
```

Thank you.

**Comment by DavidArno** [comment]
Does C# support ref properties now? Think I missed that one 😊

**Reply by HaloFour** [reply]
```cs
using System;

var c = new C();
c.Value = 5;
Console.WriteLine(c.Value);

class C {
    private int value;

    public ref int Value { get => ref value; }
}
```

**Reply by DavidArno** [reply]
When was that added?

**Reply by HaloFour** [reply]
I think the same time ref returns were?

**Reply by kyoyama-kazusa** [reply]
C# 7.x ... I've forgot the minor version of the feature born. I just know that C# 7 introduced 20+ features, and  one of the features is ref return (properties/indexers/methods).

**Reply by DavidArno** [reply]
Well, you learn something every day. Apologies for distracting away from the request. Thanks, both.

**Reply by FaustVX** [reply]
You probably already used it, `Span<>` uses `ref` property.

**Comment by jnm2** [comment, marked as answer]
Since every line in your example should unconditionally succeed, I'd turn this around and say that having [nominal deconstruction](https://github.com/dotnet/csharplang/discussions/3546) with `var` would be the ideal way to approach the first:

```cs
var {
    RefProperty: ref r,
    RefProperty2: ref readonly s,
    RefReadOnlyProperty: ref readonly t,
} = expr;
```

The next would be list destructuring which hasn't been discussed yet that I'm aware of, but which keeps coming to my mind as we work on collection literals (@CyrusNajmabadi):

```cs
var [ref a, ref readonly b, ..] = span;
```

The last is already possible in the ideal (unconditional) form today:

```cs
ref var r = ref Method();
```

**Reply by kyoyama-kazusa** [reply]
To be honest there may be some other examples to be used, but I have no idea what are they... Because `ref` variable definition is special definition than non-`ref` one; however sometimes I will use some expressions with `ref` return. Pattern matching cannot create `ref var` pattern.

Just thinking further, I may use this to check nullability such as

```csharp
ref var r = ref obj.Property;
if (!Unsafe.IsNullRef(ref r))
    // ...
```

I cannot write pattern matching like `obj.Property is not null` because I want to check its reference of the property `Property`. If we can define `ref null` as equivalent to `Unsafe.NullRef<T>()`, can we define the pattern matching like

```csharp
if (obj.Property is ref var and not ref null)
    // ...
```

?

I don't know whether this example is reasonable to be introduced here. I just want to describe a way to use `ref var` pattern. 😄

**Reply by jnm2** [reply]
I don't know either. Every time I see the topic come up of checking refs to see if they are `ref null`, the sentiment seems to be `ref null` is so surprising in the .NET ecosystem that making it mainstream is a risky move. Without having thought deeply, it seems like NRT-style analysis could make this less risky. (Would we need a full-fledged `ref?` keyword? New NullableContextAttribute encoding scheme? That sounds expensive to design and implement.)

If you started a separate discussion to suggest the pattern `ref null` which could be applied to refs, I'd be interested to see what the response would be. It makes sense to me. But I also see how it is currently considered niche.

**Reply by kyoyama-kazusa** [reply]
Thank you so much!
